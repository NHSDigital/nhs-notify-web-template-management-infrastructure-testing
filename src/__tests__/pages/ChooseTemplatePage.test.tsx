'use client';
import { useFormState } from 'react-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import ChooseTemplatePage from '../../app/choose-template/page';
import { mockDeep } from 'jest-mock-extended';

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  useFormState: jest.fn(),
}));

describe('Choose template page', () => {
  it('selects one radio button at a time', () => {
    (useFormState as jest.Mock).mockReturnValue([
      {
        formErrors: [],
        fieldErrors: {},
      },
      '/action',
      false,
    ]);
    const container = render(<ChooseTemplatePage />);
    expect(container.asFragment()).toMatchSnapshot();

    const radioButtons = [
      screen.getByTestId('email-radio'),
      screen.getByTestId('nhs-app-radio'),
      screen.getByTestId('sms-radio'),
      screen.getByTestId('letter-radio'),
    ];
    const submitButton = screen.getByTestId('submit-button');

    for (const radioButton of radioButtons) {
      expect(radioButton).toBeInTheDocument();
      expect(radioButton).not.toBeChecked();
    }
    expect(submitButton).toBeInTheDocument();

    for (const [index, radioButton] of radioButtons.entries()) {
      // select an option
      fireEvent(radioButton, new MouseEvent('click'));

      // make sure the selected option is checked and the other options are not
      for (const [index2, radioButton2] of radioButtons.entries()) {
        if (index === index2) {
          expect(radioButton2).toBeChecked();
        } else {
          expect(radioButton2).not.toBeChecked();
        }
      }
    }
  });

  it('renders error component', () => {
    (useFormState as jest.Mock).mockReturnValue([
      {
        formErrors: [],
        fieldErrors: {
          'choose-template': ['Component error message'],
        },
      },
      '/action',
      false,
    ]);
    const container = render(<ChooseTemplatePage />);
    expect(container.asFragment()).toMatchSnapshot();

    expect(screen.getByTestId('error-summary')).toBeInTheDocument();
    expect(
      document.getElementById('choose-template-error-message')
    ).toBeInTheDocument();
  });
});
