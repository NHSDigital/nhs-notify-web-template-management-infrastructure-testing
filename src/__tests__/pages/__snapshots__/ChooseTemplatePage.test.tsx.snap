// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choose template page renders error component 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-error-summary"
  >
    <h2
      class="nhsuk-error-summary__title"
      data-testid="error-summary"
    >
      There is a problem
    </h2>
    <ul
      class="nhsuk-list nhsuk-error-summary__list"
    >
      <li>
        <a
          href="#choose-template"
        >
          Component error message
        </a>
      </li>
    </ul>
  </div>
  <form
    action="/action"
  >
    <div
      class="nhsuk-form-group nhsuk-form-group--error"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--l"
        >
          <h1
            class="nhsuk-fieldset__heading"
          >
            Choose a template type to create
          </h1>
        </legend>
        <div
          class="nhsuk-form-group nhsuk-form-group--error"
        >
          <span
            class="nhsuk-error-message"
            id="choose-template-error-message"
          >
            <span
              class="nhsuk-u-visually-hidden"
            >
              Error: 
            </span>
            Component error message
          </span>
          <div
            class="nhsuk-radios"
            id="choose-template"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-1--label"
                class="nhsuk-radios__input"
                data-testid="nhs-app-radio"
                id="choose-template-1"
                name="choose-template"
                type="radio"
                value="nhs-app"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-1"
                id="choose-template-1--label"
              >
                NHS App message
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-2--label"
                class="nhsuk-radios__input"
                data-testid="email-radio"
                id="choose-template-2"
                name="choose-template"
                type="radio"
                value="email"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-2"
                id="choose-template-2--label"
              >
                Email
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-3--label"
                class="nhsuk-radios__input"
                data-testid="sms-radio"
                id="choose-template-3"
                name="choose-template"
                type="radio"
                value="sms"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-3"
                id="choose-template-3--label"
              >
                Text message (SMS)
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-4--label"
                class="nhsuk-radios__input"
                data-testid="letter-radio"
                id="choose-template-4"
                name="choose-template"
                type="radio"
                value="letter"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-4"
                id="choose-template-4--label"
              >
                Letter
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;

exports[`Choose template page selects one radio button at a time 1`] = `
<DocumentFragment>
  <form
    action="/action"
  >
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--l"
        >
          <h1
            class="nhsuk-fieldset__heading"
          >
            Choose a template type to create
          </h1>
        </legend>
        <div
          class="nhsuk-form-group"
        >
          <div
            class="nhsuk-radios"
            id="choose-template"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-1--label"
                class="nhsuk-radios__input"
                data-testid="nhs-app-radio"
                id="choose-template-1"
                name="choose-template"
                type="radio"
                value="nhs-app"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-1"
                id="choose-template-1--label"
              >
                NHS App message
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-2--label"
                class="nhsuk-radios__input"
                data-testid="email-radio"
                id="choose-template-2"
                name="choose-template"
                type="radio"
                value="email"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-2"
                id="choose-template-2--label"
              >
                Email
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-3--label"
                class="nhsuk-radios__input"
                data-testid="sms-radio"
                id="choose-template-3"
                name="choose-template"
                type="radio"
                value="sms"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-3"
                id="choose-template-3--label"
              >
                Text message (SMS)
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="choose-template-4--label"
                class="nhsuk-radios__input"
                data-testid="letter-radio"
                id="choose-template-4"
                name="choose-template"
                type="radio"
                value="letter"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="choose-template-4"
                id="choose-template-4--label"
              >
                Letter
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;
